design/230728_tinkerbell_workflow_execution.md

```diff
@@ -0,0 +1,491 @@
+
+The Cancelling state separates Workflows we intend to cancel from those that have been cancelled. This allows Tink Server to issue a cancellation instruction to Tink Agent, and for Tink Agent to indicate successful cancellation. The explicitness of this sequence is useful because:
+* Explicit indication that a Running Workflow has stopped and thus Tink Agent is available for a subsequent Workflow is important to avoid unintended Workflow rejection (double dispatch).
+* It aids in avoiding a double dispatch. Canceled is an end state so the system is at liberty to dispatch a subsequent Workflow even if the previous Workflow is still in the process of cancelling.
+* It makes clear to operators the state of the Workflow and why Tink Agent isn't executing a subsequent Workflow.
+
+### Executing Workflows
+
+Tink Server will continue to expose a gRPC API. The protobuf will be simplified to remove superfluous APIs and data structures. Tink Agent will connect to Tink Server to create a command stream. When Tink Server identifies a runnable Workflow for a connected Tink Agent it will dispatch the Workflow on the command stream. Workflows are dispatched in the order they are created. As Tink Agent executes Workflow Actions it will publish events to Tink Server.
+
+<p align="center">
+  <img src="images/230728_tinkerbell_workflow_execution/executing-a-workflow.png" />
+</p>
+
+#### Service protobuf
+
+The Workflow Service exposes 2 APIs, 1 for establishing a command stream and 1 for publishing Workflow execution events. The full protobuf can be found in Appendix A.
+
+```protobuf
+// WorkflowService is responsible for retrieving workflows to be executed by the agent and
+// publishing events as a workflow executes.
+service WorkflowService {
+  // GetWorkflows creates a stream that will receive workflows intended for the agent identified
+  // by the GetWorkflowsRequest.agent_id.
+  rpc GetWorkflows(GetWorkflowsRequest) returns (stream GetWorkflowsResponse) {}
+
+  // PublishEvent publishes a workflow event.
+  rpc PublishEvent(PublishEventRequest) returns (PublishEventResponse) {}
+}
+```
+
+#### Workflow Events
+
+| Event               | Description                                                                                     |
+| ------------------- | ----------------------------------------------------------------------------------------------- |
+| `ActionStarted`     | The action has begun executing                                                                  |
+| `ActionSucceeded`   | The action has completed successfully                                                           |
+| `ActionFailed`      | The action failed to complete. Appropriate failure reason and message will accompany the event  |
+| `WorkflowRejected`  | Tink Agent has explicitly rejected the request to run the Workflow |
+
+### Cancelling a Workflow
+
+Operators can cancel Workflows by deleting the Workflow or setting the status to Cancelling.
```

Comment:
Something about having the user tamper with the `Status` field is wrong. An alternative:

1. Introduce a `Cancel bool` field to Workflows
2. `Cancel` can be toggled on by Tink Controller when the Workflow is deleted. 
3. Tink Server will receive the update and cancel the Workflow. When it dispatches the cancellation command to Tink Agent it can transition the Workflow to Cancelling. 
4. The Workflow is transitioned to Cancelled either when an end state event is received by Tink Server or if the timeout trips and Tink Controller forces it into a cancelled state.

Reply:
i like this. avoiding the modification of status feel appropriate.

Reply:
@jacobweinstock and I brainstormed another option.

We can denote in documentation that users must populate an arbitrary finalizer to prevent the resource from being deleted, which is what a specific `Cancel` field would really accommodate, before deleting the Workflow. This makes deleting a Workflow a 2 step process. When we develop the CLI component for Tinkerbell we can add a command to delete Workflows with a parameter to keep the resource around.